### DISCUSSION TITLE
[@types/react]: TS error when returning undefined from forwardRef render function

### DISCUSSION THREAD

**Discussion by Fazla-GroM** [DefinitelyTyped/DefinitelyTyped#64800] [Issues with a @types package]
So from React@18 we can return `undefined` from a component:
```tsx
// This will render, no TS errors...
const MyComponent = ({ condition }: { condition: boolean }) => {
    if (s) {
        return <div></div>
    }
}
```
But with forward ref TS will [complain](https://ts-error-translator.vercel.app/?error=IIJw5grgtgpgdgFwAQHsBmSEE8AOMkDkAFAN5IDGKcAJgJYK1UA0SAdOyDAM7IC+AXEgDyUegB4AIjAQBDWgBsY1ABIAVALIAZAAogUOLmIBCEBAipqtwMyFoAjU9zGXNJs1QCii2IgB8LFzdzOC8YHwR-JAAiTjQo3yQAMiQydlYAbiReFliYTjhyGEEAMRQQAHcZEGolACUYNDE03wBKJABeBNDwpAAfJAgaBto4JQIkWi4kOBRkGS4uWjA4GTtFTBQkHCqZWAQ81AxsPEJSiqrqerR6oZBiwfIGKmcNV1Ng7vgEFhFxKVkFEoXLp9IYghZXtYELYHPtDIF3p5vF9Igj3CFkX4WDEGvEkik2OxMrxfARWAAoJBIVS4fAET6IPoDIZoEZjCZTGZzBZLFZrfDmTC0wj1GSPBkIMQyOBYFg8WxwMBMgBSAGUABoSgDCVHlEEeZSlMt8CX6cAg8nkZMpVOpwoIgxqrNG1HGk2msyQ80Wy1W60FxzpovFmMl0tlSHlIyV-TVmtDOrgeoNICNWBNTPNlrJQA):
```tsx
const MyComponent = forwardRef<HTMLButtonElement, React.ComponentPropsWithoutRef<'button'> & { condition: boolean }>(
    ({ condition, ...rest }, reference) => {
        if (condition) {
            return <button {...rest} ref={reference}></button>
        }
    }
)
```
I understand that I could return `null` and be done with it, but in the end it all comes to personal preferences and code styles. 

Fix seems simple (althought i dont know if it will break some other types), we just need to add missing `undefined`:
```ts
 interface ForwardRefRenderFunction<T, P = {}> {
        // extend return type
        (props: P, ref: ForwardedRef<T>): ReactElement | null | undefined;
        displayName?: string | undefined;
        // explicit rejected with `never` required due to
        // https://github.com/microsoft/TypeScript/issues/36826
        /**
         * defaultProps are not supported on render functions
         */
        defaultProps?: never | undefined;
        /**
         * propTypes are not supported on render functions
         */
        propTypes?: never | undefined;
    }
```

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
This is mostly for type-parity for standard function components. For those we can't have proper types due to https://github.com/microsoft/TypeScript/issues/21699 which isn't a `@types/react` but `typescript` issue.

We could make an argument that we could fix this issue just for `forwardRef`. But that would make it inconsistent and create refactoring hazards where you couldn't easily remove the `forwardRef` without breaking return types.
